qiskit/circuit/singleton.py

```diff
@@ -0,0 +1,250 @@
+# applied, and has an `__init_subclass__` that applies the singleton-subclass-creation steps above.
+# The overrides are in separate classes so that _mutable_ `XGate` instances do not have them in
+# their own MROs; doing this is easier to implement, but requires all the setters and checkers to
+# dance around at runtime trying to validate whether mutating the instance is allowed.
+
+
+def _impl_new(cls, *_args, **_kwargs):
+    # __new__ for the singleton instances.
+    raise TypeError(f"cannot create '{cls.__name__}' instances")
+
+
+def _impl_init_subclass(base, overrides):
+    # __init_subclass__ for the classes that make their children singletons (e.g. `SingletonGate`)
+
+    def __init_subclass__(cls, *, create_singleton=True, **kwargs):
+        super(base, cls).__init_subclass__(**kwargs)
+        if not create_singleton:
+            return
+
+        # We need to make a new type object that pokes in the overrides into the correct
+        # place in the method-resolution order.
+        singleton_class = _SingletonMeta.__new__(
+            _SingletonMeta,
+            f"_Singleton{cls.__name__}",
+            (cls, overrides),
+            # This is a dynamically generated class so it's got no module.  The slot layout of the
+            # singleton class needs to match any layout in the base.
+            {"__module__": None, "__slots__": (), "__new__": _impl_new, "_base_class": cls},
+            create_singleton=False,
+        )
+
+        # Make a mutable instance, fully instantiate all lazy properties, then freeze it.
+        cls._singleton_instance = cls(_force_mutable=True)
+        cls._singleton_instance._define()
+        cls._singleton_instance.__class__ = singleton_class
+
+    return classmethod(__init_subclass__)
+
+
+class _SingletonMeta(type(Instruction)):
+    # The inheritance above is to ensure metaclass compatibility with `Instruction`, though pylint
+    # doesn't understand that this is a subclass of `type`, so uses metaclass `self` conventions.
+
+    # pylint: disable=bad-classmethod-argument,no-self-argument
+
+    __slots__ = ()
+
+    def __new__(mcs, name, bases, namespace, *, overrides=None, **kwargs):
+        cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+        if overrides is not None:
+            # The `__init_subclass__` logic is shared between `SingletonInstruction` and
+            # `SingletonGate`, but we can't do `__init_subclass__ = _impl_init_subclass(...)`
+            # inside the class definition bodies because it couldn't make `super()` calls.
+            cls.__init_subclass__ = _impl_init_subclass(cls, overrides)
+        return cls
+
+    def __call__(cls, *args, _force_mutable=False, **kwargs):
+        if not _force_mutable and not args and not kwargs:
+            # This class attribute is created by the singleton-creation base classes'
+            # `__init_subclass__` methods; see `_impl_init_subclass`.
+            return cls._singleton_instance
```

Comment:
My memory of the ordering here was a bit rusty,  I was originally thinking that `__init_subclass__` gets called as part of `super().__call__()` in the normal class creation steps. But it must get called independently for this to work because we'd end up potentially never having set the class attribute if a mutable instance is never created so I went to read pep 487 and see how it works now. But maybe for future reference expanding the comment a little bit to say `__init_subclass__` gets invoked in this context.

Reply:
`__init_subclass__` is a hook that `type.__call__` calls in the type-creation overload after the type object has been allocated by the standard methods, but before the object gets returned.  It's give-or-take just an easier way to do a large amount of what people typically use metaclasses for.  `_SingletonMeta` is still a metaclass most because we need to override `__call__` on the type objects, which `__init_subclass__` doesn't let us do, because the magic `__call__` used to overload class construction is looked up on the _type_ of the type object given to `__init_subclass__`, not the type object itself.

It's a bit hard to explain the "type" terminology here because there are several objects that both _are_ types and yet of course still have a type of their own.

Reply:
Added a bit more comment in the class definition in 145decf53e.